VultisigApp/VultisigApp/Services/1inch/OneInchService.swift

```diff
@@ -36,26 +36,57 @@ struct OneInchService {
             fee: 0.5,
             isAffiliate: isAffiliate
         )
-
+        
         var request = URLRequest(url: url)
         request.allHTTPHeaderFields = [
             "accept": "application/json",
         ]
-
+        
         let (data, _) = try await URLSession.shared.data(for: request)
         let response = try JSONDecoder().decode(OneInchQuote.self, from: data)
         return response
     }
-
+    
     func fetchTokens(chain: Int) async throws -> [OneInchToken] {
         let response: OneInchTokensResponse = try await Utils.fetchObject(from: Endpoint.fetchTokens(chain: chain))
         let tokens = Array(arrayLiteral: response.tokens.values).reduce([], +)
         return tokens
     }
+    
+    func fetchTokensBalance(chainId: Int, walletAddress: String) async throws -> [OneInchToken] {
+        let oneInchChainId = String(describing: chainId)
+        let balanceData = try await Utils.asyncGetRequest(urlString: Endpoint.fetch1InchsTokensBalance(chain: oneInchChainId, address: walletAddress), headers: [:])
+        
+        if let balanceDictionary = try JSONSerialization.jsonObject(with: balanceData, options: []) as? [String: String] {
+            
+            let tokensWithNonZeroBalance = balanceDictionary.filter { tokenAddress, balance in
+                if let balanceValue = Int64(balance) {
+                    return balanceValue > 0
+                }
+                return false
+            }
+            
+            let nonZeroBalanceTokenAddresses = tokensWithNonZeroBalance.map { $0.key }
+            
+            let tokenInfoData = try await Utils.asyncGetRequest(
+                urlString: Endpoint.fetch1InchsTokensInfo(chain: oneInchChainId, addresses: nonZeroBalanceTokenAddresses), headers: [:]
+            )
+            
+            var tokens: [OneInchToken] = []
+            for tokenAddress in nonZeroBalanceTokenAddresses {
+                if let tokenInfo: OneInchToken = Utils.extractResultFromJson(fromData: tokenInfoData, path: tokenAddress, type: OneInchToken.self) {
```

Comment:
Can you parse in array? [OneInchToken].self
But better to use JSONDecoder imo

Reply:
in this case, it is not possible, because it is not an array returned from 1inch, each address is the key, then the value is the ```OneInchToken```

```
{
  "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee": {
    "address": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
    "chainId": 1,
    "decimals": 18,
    "name": "Ether",
    "symbol": "ETH",
    "providers": [
      "1inch",
      "Curve Token List"
    ],
    "logoURI": "https://tokens.1inch.io/0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee.png",
    "eip2612": false,
    "tags": [
      "native",
      "PEG:ETH"
    ],
    "rating": 2
  },
  "0x85f17cf997934a597031b2e18a9ab6ebd4b9f6a4": {
    "address": "0x85f17cf997934a597031b2e18a9ab6ebd4b9f6a4",
    "chainId": 1,
    "decimals": 24,
    "name": "NEAR Protocol",
    "symbol": "NEAR",
    "providers": [
      "1inch",
      "CoinGecko",
      "Kleros Tokens"
    ],
    "logoURI": "https://tokens.1inch.io/0x85f17cf997934a597031b2e18a9ab6ebd4b9f6a4.png",
    "eip2612": false,
    "tags": [
      "tokens"
    ],
    "rating": 3
  }
}
```